src/goto-programs/goto_contractor.cpp

```diff
@@ -0,0 +1,453 @@
+      map->intervals[i].lb() != new_intervals[i].lb())
+    {
+      auto lb = create_signed_32_value_expr(new_intervals[i].lb());
+      auto cond = create_greaterthanequal_relation(X, lb);
+      goto_programt tmp_e(message_handler);
+      goto_programt::targett e = tmp_e.add_instruction(ASSUME);
+      e->inductive_step_instruction = false;
+      e->guard = cond;
+      e->location = loop_exit->location;
+      goto_function.body.destructive_insert(loop_exit, tmp_e);
+    }
+    if(
+      isfinite(new_intervals[i].ub()) &&
+      map->intervals[i].ub() != new_intervals[i].ub())
+    {
+      auto ub = create_signed_32_value_expr(new_intervals[i].ub());
+      auto cond = create_lessthanequal_relation(X, ub);
+      goto_programt tmp_e(message_handler);
+      goto_programt::targett e = tmp_e.add_instruction(ASSUME);
+      e->inductive_step_instruction = false;
+      e->guard = cond;
+      e->location = loop_exit->location;
+      goto_function.body.destructive_insert(loop_exit, tmp_e);
+    }
+  }
+}
+
+IntervalVector goto_contractort::contractor()
+{
+  NumConstraint c = *constraint;
+
+  std::ostringstream oss;
+  oss << "\t- Constraint:" << c;
+  oss << "\n\t- Function:" << c.f;
+
+  auto complement = get_complement(c.op);
+
+  NumConstraint c2(c.f, complement);
+  oss << "\n\t- Complement: " << c2;
+
+  CtcFwdBwd c_out(c);
+  CtcFwdBwd c_in(c2);
+
+  domains = map->intervals;
+  oss << "\n\t- Domains (before): " << domains;
+  auto X = domains;
+
+  c_in.contract(X);
+
+  ///Keep this for later
+  IntervalVector *s_in;
+  //int num = domains.diff(X, s_in);
+  oss << "\n\t- Domains (after): " << X;
+  //std::cout << "My domains after Inner contractor:" << X << std::endl;
+  //  for(int i = 0; i < num; i++)
+  //    oss << "\n\t- "
+  //    message_handler.debug( "s_in[%d]: %f",i,  s_in[i]));
+
+  message_handler.status(oss.str());
+
+  return X;
+}
+
+ibex::CmpOp goto_contractort::get_complement(ibex::CmpOp op)
+{
+  switch(op)
+  {
+  case GEQ:
+    return LT;
+  case GT:
+    return LEQ;
+  case LEQ:
+    return GT;
+  case LT:
+    return GEQ;
+  default:
+    message_handler.status("cant process equal");
+    //abort();
+    break;
+  }
+  return GEQ;
+}
+
+Ctc *goto_contractort::create_contractors_from_expr2t(irep_container<expr2t>)
+{
+  return nullptr;
+}
+
+NumConstraint *
+goto_contractort::create_constraint_from_expr2t(irep_container<expr2t> expr)
+{
+  NumConstraint *c = nullptr;
+  if(
+    is_arith_expr(expr) || is_constant_number(expr) || is_symbol2t(expr) ||
+    is_notequal2t(expr))
+  {
+    message_handler.status("Expression is complex, skipping this assert");
+  }
+
+  std::shared_ptr<relation_data> rel;
+  rel = std::dynamic_pointer_cast<relation_data>(expr);
+
+  Function *f, *g;
+  f = create_function_from_expr2t(rel->side_1);
+  g = create_function_from_expr2t(rel->side_2);
+  switch(expr->expr_id)
+  {
+  case expr2t::expr_ids::greaterthanequal_id:
+    c = new NumConstraint(*vars, (*f)(*vars) >= (*g)(*vars));
```

Comment:
do we call delete on these object anywhere?

Reply:
Created destructor.